File: ZTTextField-Objc/ZTTextFieldText/ZTTextField/ZTTextField.m
//
//  ZTTextField.m
//  ZTTextFieldText
//
//  Created by zhaitong on 2017/12/14.
//  Copyright © 2017年 zhaitong. All rights reserved.
//

#import "ZTTextField.h"
@class LRTextField;

static CGFloat const placeholderLabelShowScale = 0.8978;
static CGFloat const placeholderLabelHideScale = 1.02;

@interface ZTTextField () <CAAnimationDelegate>
@property (nonatomic) UILabel *placeholderLabel;
@property (nonatomic) UILabel *hintLabel;
@property (nonatomic, copy) NSString *sWrongFormatMsg;
@end

@implementation ZTTextField

- (void)awakeFromNib {
    [super awakeFromNib];
    [self updateUI];
}

- (void)updateUI {
    self.borderStyle = UITextBorderStyleNone;
    self.clipsToBounds = NO;
    self.backgroundColor = [UIColor clearColor];
    [self createUpperPlaceholderLabel];
    [self createHintLabel];
   
    [self addTarget:self action:@selector(textFieldEdittingDidBeginInternal:) forControlEvents:UIControlEventEditingDidBegin];
    [self addTarget:self action:@selector(textFieldEdittingDidChangeInternal:) forControlEvents:UIControlEventEditingChanged];
    [self addTarget:self action:@selector(textFieldEdittingDidEndInternal:) forControlEvents:UIControlEventEditingDidEnd];
}

- (void)setValidationBlk:(FormatValidationBlock)validationBlk {
    _validationBlk = validationBlk;
    if (validationBlk) {
        self.sWrongFormatMsg = _validationBlk(self);
    }
}

- (void)createUpperPlaceholderLabel {
    self.placeholderLabel = [UILabel new];
    self.placeholderLabel.text = self.editingPhText;
    self.placeholderLabel.font = [UIFont systemFontOfSize:self.upperPhFontSize];//展示在👆的placeholder的字体大小
    [self.placeholderLabel sizeToFit];
    [self floatingLabelUpperColor];
    if (self.text.length > 0) {
        self.placeholderLabel.transform = CGAffineTransformMakeScale(placeholderLabelShowScale, placeholderLabelShowScale);//缩放
        self.placeholderLabel.frame = [self floatingLabelUpperFrame];
    } else {
        self.placeholderLabel.transform = CGAffineTransformMakeScale(placeholderLabelHideScale, placeholderLabelHideScale);//还原
        self.placeholderLabel.frame = self.textFieldType == ZTTextFieldTypeNormal ? [super textRectForBounds:self.bounds] : [self noBorderFloatingLabelFrame];
    }
    [self addSubview:self.placeholderLabel];
}

- (void)createHintLabel {
    self.hintLabel = [UILabel new];
    self.hintLabel.transform = CGAffineTransformMakeScale(placeholderLabelShowScale, placeholderLabelShowScale);//缩放
    self.hintLabel.font = [UIFont systemFontOfSize:self.hintLabelFontSize];
    self.hintLabel.textAlignment = NSTextAlignmentRight;
    self.hintLabel.textColor = self.placeholderInactiveColor;
    [self addSubview:self.hintLabel];
}



- (void)showBeginAnimation {
    [UIView transitionWithView:self.placeholderLabel
                      duration:0.35f
                       options:UIViewAnimationOptionBeginFromCurrentState | UIViewAnimationOptionTransitionFlipFromTop
                    animations:^{
                        self.placeholderLabel.text = self.upperPhText;
                        [self.placeholderLabel sizeToFit];
                        [self floatingLabelUpperColor];
                        self.placeholderLabel.transform = CGAffineTransformMakeScale(placeholderLabelShowScale, placeholderLabelShowScale);//缩放
                        self.placeholderLabel.frame = [self floatingLabelUpperFrame];
                    } completion:nil];
    [UIView transitionWithView:self.hintLabel
                      duration:0.35f
                       options:UIViewAnimationOptionBeginFromCurrentState | UIViewAnimationOptionTransitionCrossDissolve
                    animations:^{
                        self.hintLabel.alpha = 1;
                        self.hintLabel.text = self.hintLabelText;
                        [self.hintLabel sizeToFit];
                        self.hintLabel.textColor = self.placeholderInactiveColor;
                        [self hintLabelFrame];
                    } completion:nil];
 
}

- (void)showEndAnimation {
    [UIView transitionWithView:self.placeholderLabel
                      duration:0.35f
                       options:UIViewAnimationOptionBeginFromCurrentState | UIViewAnimationOptionTransitionFlipFromBottom
                    animations:^{
                        self.placeholderLabel.text = self.editingPhText;
                        [self.placeholderLabel sizeToFit];
                        [self floatingLabelUpperColor];
                        self.placeholderLabel.transform = CGAffineTransformMakeScale(placeholderLabelHideScale, placeholderLabelHideScale);//还原
                        self.placeholderLabel.frame = self.textFieldType == ZTTextFieldTypeNormal ? [super textRectForBounds:self.bounds] : [self noBorderFloatingLabelFrame];
                    } completion:nil];
    
}

- (void)shakeHintLabel:(NSString *)sWrongMsg {
    self.hintLabel.alpha = 1;
    self.hintLabel.textColor = [UIColor redColor];
    self.hintLabel.text = sWrongMsg;
    [self.hintLabel sizeToFit];
    [self hintLabelFrame];
    CABasicAnimation *animation = [CABasicAnimation animation];
    animation.delegate = self;
    [animation setDuration:0.1];
    CGFloat x = CGRectGetMaxX(self.bounds) - self.hintLabel.frame.size.width - 5 + self.hintLabel.frame.size.width / 2;
    animation.fromValue = [NSValue valueWithCGPoint:CGPointMake(x  - 10, - 1 - self.frame.size.height / 2 + self.hintLabel.frame.size.height / 2)];
    animation.toValue = [NSValue valueWithCGPoint:CGPointMake(x  + 10, - 1 - self.frame.size.height / 2 + self.hintLabel.frame.size.height / 2)];
    animation.repeatCount = 3;
    animation.autoreverses = YES;
    [self.hintLabel.layer addAnimation:animation forKey:@"position"];
}

- (void)floatingLabelUpperColor {
    self.placeholderLabel.textColor = self.editing ? self.placeholderActiveColor : self.placeholderInactiveColor;
}
- (void)hintLabelFrame {
    CGRect placeHolderFrame = self.bounds;
    CGRect hintLabelFrame = self.hintLabel.frame;
    hintLabelFrame.origin.x = CGRectGetMaxX(placeHolderFrame) - self.hintLabel.frame.size.width - 5;
    hintLabelFrame.origin.y = - 1 - self.frame.size.height / 2;
    self.hintLabel.frame = hintLabelFrame;
}

- (CGRect)floatingLabelUpperFrame {
    CGRect placeHolderFrame = [super textRectForBounds:self.bounds];
    return CGRectMake(placeHolderFrame.origin.x - 1, - 1 - self.frame.size.height / 2, self.bounds.size.width - 2, self.frame.size.height / 2);
}

- (CGRect)noBorderFloatingLabelFrame {
    CGRect placeHolderFrame = [super textRectForBounds:self.bounds];
    //颜色置灰
    return CGRectMake(placeHolderFrame.origin.x - 1,0 , self.bounds.size.width - 2, self.frame.size.height / 2);
}

- (void)validateText {

    if (self.validationBlk) {
        self.sWrongFormatMsg = self.validationBlk(self);
    }
    if (self.sWrongFormatMsg.length > 0) {
        [UIView transitionWithView:self.hintLabel
                          duration:0.35f
                           options:UIViewAnimationOptionBeginFromCurrentState | UIViewAnimationOptionTransitionCrossDissolve
                        animations:^{
                            self.hintLabel.text = self.text.length > 0 ? self.sWrongFormatMsg : self.hintLabelText;
                            [self.hintLabel sizeToFit];
                            [self hintLabelFrame];
                            self.hintLabel.textColor = self.text.length > 0 ? [UIColor redColor] : _placeholderInactiveColor;
                            self.hintLabel.alpha = self.text.length > 0 ? 1 : 0;
                        } completion:nil];
    } else {
        [UIView transitionWithView:self.hintLabel
                          duration:0.35f
                           options:UIViewAnimationOptionBeginFromCurrentState | UIViewAnimationOptionTransitionCrossDissolve
                        animations:^{
                            self.hintLabel.alpha = 0;
                        } completion:nil];
    }
}

- (void)textFieldShowWrongMessage:(NSString *)sWrongMessage {
    if (sWrongMessage.length > 0) {
        [self shakeHintLabel:sWrongMessage];
    }
}

#pragma mark - setter
- (void)setText:(NSString *)text {
    [super setText:text];
    if (text.length > 0) {
        [self showBeginAnimation];
    }
}

- (void)setTextFieldType:(ZTTextFieldType)textFieldType {
    _textFieldType = textFieldType;
    switch (textFieldType) {
        case ZTTextFieldTypeNoBorder:
            self.borderStyle = UITextBorderStyleNone;
            break;
        default:
            self.borderStyle = UITextBorderStyleRoundedRect;
            break;
    }
    self.placeholderLabel.frame = self.textFieldType == ZTTextFieldTypeNormal ? [super textRectForBounds:self.bounds] : [self noBorderFloatingLabelFrame];
}
#pragma mark - Target Method

- (IBAction)textFieldEdittingDidBeginInternal:(UITextField *)sender {
 //开始编辑
    [self runDidBeginAnimation];
}

- (IBAction)textFieldEdittingDidEndInternal:(UITextField *)sender {
    //验证文本规则
    [self validateText];
    //结束编辑
    [self runDidEndAnimation];
}
- (IBAction)textFieldEdittingDidChangeInternal:(UITextField *)sender {
    [self runDidChange];
}

/**
 * 开始编辑
 */
- (void)runDidBeginAnimation {
    
    if (self.text.length > 0 || self.editing) {//已经有内容
        [self showBeginAnimation];
    }
}

/**< 结束编辑*/
- (void)runDidEndAnimation {
    if (self.text.length > 0) {
        //执行动画
        [UIView transitionWithView:self.placeholderLabel
                          duration:0.3f
                           options:UIViewAnimationOptionBeginFromCurrentState | UIViewAnimationOptionTransitionCrossDissolve
                        animations:^{
                            self.placeholderLabel.textColor = self.placeholderInactiveColor;
                        } completion:nil];

    } else {
        
        [self showEndAnimation];
    }
}
/**< 正在编辑*/
- (void)runDidChange {
    if (self.text.length == 0) {
        
    }
}
- (void)animationDidStop:(CAAnimation *)anim finished:(BOOL)flag {
    [UIView transitionWithView:self.hintLabel
                      duration:1.5
                       options:UIViewAnimationOptionBeginFromCurrentState | UIViewAnimationOptionTransitionCrossDissolve
                    animations:^{
                        self.hintLabel.alpha = 0;
                    } completion:nil];
}
  
@end

File: ZTTextField-Swift/ZTTextField-Swift/ZTTextField/ZTTextField.swift
//
//  ZTTextField.swift
//  ZTTextField-Swift
//
//  Created by zhaitong on 2017/12/22.
//  Copyright © 2017年 zhaitong. All rights reserved.
//

import UIKit


enum ZTTextFieldType {
    case ZTTextFieldTypeNoBorder
    case ZTTextFieldTypeNormal
}

typealias FormatValidationClosure = (_ textField : ZTTextField) -> String

@IBDesignable

class ZTTextField: UITextField,CAAnimationDelegate {
    static let placeholderLabelShowScale : CGFloat = 0.8978
    static let placeholderLabelHideScale : CGFloat = 1.02
    
    @IBInspectable fileprivate var upperPhText : String? //Text to be displayed above the field.
    @IBInspectable fileprivate var editingPhText : String? //Text to be displayed in the field.
    @IBInspectable fileprivate var upperPhFontSize : CGFloat //FontSize of placeholder Label.
    @IBInspectable fileprivate var placeholderInactiveColor : UIColor? //Text color to be applied to floating placeholder text when editing.
    /**
     * Text color to be applied to floating placeholder text when editing.
     * Default is tint color.
     */
    @IBInspectable fileprivate var placeholderActiveColor : UIColor?
    /**
     * Hint label text.
     */
    @IBInspectable fileprivate var hintLabelText : String?
    /**
     * FontSize of hint Label.
     */
    @IBInspectable fileprivate var hintLabelFontSize : CGFloat
    /**
     * block vertify the text format
     */
    open var validationCloure : FormatValidationClosure? {
        
        willSet {
            if let tempValue = newValue {
                self.sWrongFormatMsg = tempValue(self)
            }
        }
    }
    /**
     * The type of the textfield
     */
    open var textFieldType : ZTTextFieldType = .ZTTextFieldTypeNoBorder{
        didSet {
            switch textFieldType {
            case .ZTTextFieldTypeNoBorder:
                self.borderStyle = UITextBorderStyle.none
            default:
                self.borderStyle = UITextBorderStyle.roundedRect
            }
            self.placeholderLabel?.frame = self.textFieldType == .ZTTextFieldTypeNormal ? super.textRect(forBounds: self.bounds) : self.noBorderFloatingLabelFrame()
        }
    }
    private var placeholderLabel : UILabel?
    private var hintLabel : UILabel?;
    private var sWrongFormatMsg : String?;
    
    override open var text: String? {
        didSet {
            super.text = text
            if let textCount = text?.count {
                if textCount > 0 {
                    self.showBeginAnimation()
                }
            }
        }
    }
    
    required public init?(coder aDecoder: NSCoder) {
        upperPhFontSize = 15
        hintLabelFontSize = 15
        super.init(coder: aDecoder)
        self.updateUI()
    }
    override init(frame: CGRect) {
        upperPhFontSize = 15
        hintLabelFontSize = 15
        super.init(frame: frame)
        self.updateUI()
    }
    override func awakeFromNib() {
        super.awakeFromNib()
        self.updateUI()
    }
    fileprivate final func updateUI() {
        borderStyle = .none
        clipsToBounds = false;
        backgroundColor = UIColor.clear
    
        self.createUpperPlaceholderLabel()
        self.createHintLabel()
        self.addTarget(self, action: #selector(self.textFieldEdittingDidBeginInternal), for: .editingDidBegin)
        self.addTarget(self, action: #selector(self.textFieldEdittingDidChangeInternal), for: .editingChanged)
        self.addTarget(self, action: #selector(self.textFieldEdittingDidEndInternal), for: .editingDidEnd)
    }
    func createUpperPlaceholderLabel() {
        self.placeholderLabel = UILabel.init()
        self.placeholderLabel?.text = self.editingPhText
        self.placeholderLabel?.font = UIFont.systemFont(ofSize: self.upperPhFontSize)
        self.placeholderLabel?.sizeToFit()
        self.floatingLabelUpperColor()
        if let textCount = self.text?.count {
            if textCount > 0 {
                self.placeholderLabel?.transform = CGAffineTransform(scaleX: ZTTextField.placeholderLabelShowScale, y: ZTTextField.placeholderLabelShowScale);//缩放
                self.placeholderLabel?.frame = self.floatingLabelUpperFrame()
            } else {
                self.placeholderLabel?.transform = CGAffineTransform(scaleX: ZTTextField.placeholderLabelHideScale, y: ZTTextField.placeholderLabelHideScale);//还原
                self.placeholderLabel?.frame = self.textFieldType == .ZTTextFieldTypeNormal ? super.textRect(forBounds: self.bounds) : self.noBorderFloatingLabelFrame()
            }
        }
        self.addSubview(self.placeholderLabel!)
    }
    func createHintLabel() {
        self.hintLabel = UILabel.init()
        self.hintLabel?.transform = CGAffineTransform(scaleX: ZTTextField.placeholderLabelShowScale, y: ZTTextField.placeholderLabelShowScale)//缩放
        self.hintLabel?.font = UIFont.systemFont(ofSize: self.hintLabelFontSize)
        self.hintLabel?.textAlignment = NSTextAlignment.right
        self.hintLabel?.textColor = self.placeholderInactiveColor
        self.addSubview(self.hintLabel!)
        
    }
    func floatingLabelUpperColor() {
        self.placeholderLabel?.textColor = self.isEditing ? self.placeholderActiveColor : self.placeholderInactiveColor
    }
    func showBeginAnimation() {
        UIView.transition(with: self.placeholderLabel!, duration: 0.35, options: [UIViewAnimationOptions.beginFromCurrentState,UIViewAnimationOptions.transitionFlipFromTop], animations: {
            self.placeholderLabel?.text = self.upperPhText
            self.placeholderLabel?.sizeToFit()
            self.floatingLabelUpperColor()
            self.placeholderLabel?.transform = CGAffineTransform(scaleX: ZTTextField.placeholderLabelShowScale, y: ZTTextField.placeholderLabelShowScale)//缩放
            self.placeholderLabel?.frame = self.floatingLabelUpperFrame()
        }, completion: nil)
        UIView.transition(with: self.hintLabel!, duration: 0.35, options: [UIViewAnimationOptions.beginFromCurrentState,UIViewAnimationOptions.transitionCrossDissolve], animations: {
            self.hintLabel?.alpha = 1;
            self.hintLabel?.text = self.hintLabelText;
            self.hintLabel?.sizeToFit()
            self.hintLabel?.textColor = self.placeholderInactiveColor
            self.hintLabelFrame()
        }, completion: nil)
    }
    func showEndAnimation() {
        UIView.transition(with: self.placeholderLabel!, duration: 0.35, options: [UIViewAnimationOptions.beginFromCurrentState,UIViewAnimationOptions.transitionFlipFromBottom], animations: {
            self.placeholderLabel?.text = self.editingPhText;
            self.placeholderLabel?.sizeToFit()
            self.floatingLabelUpperColor()
            self.placeholderLabel?.transform = CGAffineTransform(scaleX: ZTTextField.placeholderLabelHideScale, y: ZTTextField.placeholderLabelHideScale)//还原
            self.placeholderLabel?.frame = self.textFieldType == .ZTTextFieldTypeNormal ? super.textRect(forBounds: self.bounds) : self.noBorderFloatingLabelFrame()
        }, completion: nil)
    }
    func floatingLabelUpperFrame() -> CGRect {
        let placeHolderFrame : CGRect = super.textRect(forBounds: self.bounds)
        return CGRect(x: placeHolderFrame.origin.x - 1, y: -1 - self.frame.size.height / 2, width: self.bounds.size.width - 2, height: self.frame.size.height / 2)
    }
    func noBorderFloatingLabelFrame() -> CGRect {
        let placeHolderFrame : CGRect = super.textRect(forBounds: self.bounds)
        //颜色置灰
        return CGRect(x: placeHolderFrame.origin.x - 1, y: 0, width: self.bounds.size.width - 2, height: self.frame.size.height / 2)
    }
    func hintLabelFrame() {
        var hintLabelFrame :  CGRect = (self.hintLabel?.frame)!
        hintLabelFrame.origin.x = self.bounds.maxX - (self.hintLabel?.frame)!.size.width - 5
        hintLabelFrame.origin.y = -1 - self.frame.size.height / 2
        self.hintLabel?.frame = hintLabelFrame
    }
    func shakeHintLabel() {
        self.hintLabel?.alpha = 1
        self.hintLabel?.textColor = UIColor.red
        self.hintLabel?.text = self.sWrongFormatMsg
        self.hintLabel?.sizeToFit()
        self.hintLabelFrame()
        let animation = CABasicAnimation.init()
        animation.delegate = self
        animation.duration = 0.1
        let hintLabelX = self.bounds.maxX - (self.hintLabel?.frame.size.width)! - 5 + (self.hintLabel?.frame.size.width)! / 2
        animation.fromValue = NSValue.init(cgPoint: CGPoint.init(x: hintLabelX  - 10, y: -1 - self.frame.size.height / 2 + (self.hintLabel?.frame.size.height)! / 2))
        animation.toValue = NSValue.init(cgPoint: CGPoint.init(x: hintLabelX  + 10, y: -1 - self.frame.size.height / 2 + (self.hintLabel?.frame.size.height)! / 2))
        animation.repeatCount = 3
        animation.autoreverses = true
        self.hintLabel?.layer.add(animation, forKey: "position")
    }
    func validateText() {
        if let validationCloure = self.validationCloure {
            self.sWrongFormatMsg = validationCloure(self)
        }
        if let wrongMsgCount = self.sWrongFormatMsg?.count {
            if wrongMsgCount > 0 {
                UIView.transition(with: self.hintLabel!, duration: 0.35, options: [UIViewAnimationOptions.beginFromCurrentState,UIViewAnimationOptions.transitionCrossDissolve], animations: {
                    if let textCount = self.text?.count {
                        self.hintLabel?.text = textCount > 0 ? self.sWrongFormatMsg : self.hintLabelText
                        self.hintLabel?.textColor = textCount > 0 ? UIColor.red : self.placeholderInactiveColor
                        self.hintLabel?.alpha = textCount  > 0 ? 1 : 0
                    }
                    self.hintLabel?.sizeToFit()
                    self.hintLabelFrame()
                }, completion: nil)
            }
        }
        UIView.transition(with: self.hintLabel!, duration: 0.35, options: [UIViewAnimationOptions.beginFromCurrentState,UIViewAnimationOptions.transitionCrossDissolve], animations: {
            self.hintLabel?.alpha = 0
        }, completion: nil)
    }
    
    @objc func textFieldEdittingDidBeginInternal() {
        //开始编辑
        self.runDidBeginAnimation()
    }
    @objc func textFieldEdittingDidChangeInternal() {
        
    }
    @objc func textFieldEdittingDidEndInternal() {
        //验证文本规则
        self.validateText()
        //结束编辑
        self.runDidEndAnimation()
    }
    /**< 开始编辑*/
    func runDidBeginAnimation() {
        if let textCount = self.text?.count {
            if textCount > 0 || self.isEditing {
                self.showBeginAnimation()
            }
        }
    }
    /**< 结束编辑*/
    func runDidEndAnimation() {
        if let textCount = self.text?.count {
            if textCount > 0 {
                UIView.transition(with: self.placeholderLabel!, duration: 0.35, options: [UIViewAnimationOptions.beginFromCurrentState,UIViewAnimationOptions.transitionFlipFromTop], animations: {
                    self.placeholderLabel?.textColor = self.placeholderInactiveColor
                }, completion: nil)
            }
            self.showEndAnimation()
        }
    }
    open func textFieldShowWrongMessage(sWrongMessage : String) {
        if let wrongMessageCount = sWrongFormatMsg?.count {
            if wrongMessageCount > 0 {
                self.shakeHintLabel()
            }
        }
    }
    func animationDidStop(_ anim: CAAnimation, finished flag: Bool) {
        UIView.transition(with: self.hintLabel!, duration: 1.5, options: [UIViewAnimationOptions.beginFromCurrentState,UIViewAnimationOptions.transitionCrossDissolve], animations: {
            self.hintLabel?.alpha = 0;
        }, completion: nil)
    }
    
}
